lib/ecs.go

```diff
@@ -155,11 +166,118 @@ func GetPendingTasksCount(awsSess *session.Session, cluster string) int64 {
+
+	for _, service := range ecsServices {
+		if *service.DesiredCount > 0 {
+			// fmt.Printf("Looking at service %s, count = %v\n", *service.ServiceName, *service.DesiredCount)
+			taskDef, err := svc.DescribeTaskDefinition(&ecs.DescribeTaskDefinitionInput{
+				TaskDefinition: service.TaskDefinition,
+			})
+			if err != nil {
+				fmt.Printf("Unable to describe task definition %s\n", *service.TaskDefinition)
+				os.Exit(1)
+			}
+
+			var serviceMemory int64 = 0
+			var serviceCpu int64 = 0
+
+			for _, c := range taskDef.TaskDefinition.ContainerDefinitions {
+				// fmt.Printf("    Looking at container %s, needs %v mem and %v cpu\n", *c.Name, *c.Memory, *c.Cpu)
+				serviceMemory += *c.Memory
+				serviceCpu += *c.Cpu
+			}
+
+			if serviceMemory > largestServiceMemory {
+				largestServiceMemory = serviceMemory
+			}
+
+			if serviceCpu > largestServiceCpu {
+				largestServiceCpu = serviceCpu
+			}
+
+			memoryNeeded += serviceMemory * *service.DesiredCount
+			cpuNeeded += serviceCpu * *service.DesiredCount
+		}
 	}
 
-	return pendingTasks
-}
\ No newline at end of file
+	// Add back in the largest service memory and cpu needs to ensure there is enough extra capacity
+	// to launch another instance of the largest service for rolling updates
+	memoryNeeded += largestServiceMemory
+	cpuNeeded += largestServiceCpu
+
+	return memoryNeeded, cpuNeeded
+}
+
+func RightSizeAsgForEcsCluster(awsSess *session.Session, cluster string, atLeastServiceDesiredCount bool) error {
+	asgName := GetAsgNameForEcsCluster(awsSess, cluster)
+	if asgName == "" {
+		fmt.Println("Unable to find ASG name for ECS cluster ", cluster)
+		os.Exit(1)
+	}
+
+	fmt.Println("ASG found: ", asgName)
+
+	instanceType := GetInstanceTypeForAsg(awsSess, asgName)
+	fmt.Println("ASG uses instance type: ", instanceType)
+
+	ecsServices := ListServicesForEcsCluster(awsSess, cluster)
+	memoryNeeded, cpuNeeded := GetMemoryCpuNeededForEcsServices(awsSess, ecsServices)
+	fmt.Printf("Memory needed for all services with desired count > 0: %v, CPU needed: %v\n", memoryNeeded, cpuNeeded)
+
+	serversNeeded := HowManyServersNeededForAsg(instanceType, memoryNeeded, cpuNeeded)
+	fmt.Printf("ASG should have %v servers to fit all tasks\n", serversNeeded)
+
+	// If an ECS service has a desired count > serversNeeded, and atLeastServiceDesiredCount is true, set serversNeeded to
+	// largest ecs service desired count value
+	largestDesiredCount := GetLargestDesiredCountFromEcsServices(ecsServices)
+	if largestDesiredCount > serversNeeded && atLeastServiceDesiredCount {
+		serversNeeded = largestDesiredCount
+	}
+
+	asgDesired, asgMin, asgMax := GetAsgServerCount(awsSess, asgName)
+	fmt.Printf("ASG server count currently set to: desired = %v, min = %v, max = %v\n", asgDesired, asgMin, asgMax)
+
+	if asgMin < serversNeeded {
+		fmt.Printf("ASG needs to be scaled up by %v servers\n", serversNeeded-asgMin)
+		fmt.Printf("Scaling ASG to %v servers (desired/min/max)...", serversNeeded)
```

Comment:
Do you really want `(desired/min/max)`  in these print statements?  (here and below)

Reply:
fixed